### ISSUE TITLE
cupy.random.default_rng() leads to incorrect results for binomial distribution

### ISSUE THREAD

**Issue by Neburski** [cupy/cupy#7164]
### Description

Using the default random number generator returned by `cupy.random.default_rng()` leads to incorrect results for the binomial distribution if the input probability has a shape larger than (~540, 480) (approximate shape given here, see script below). 

For instance, in the example below, an input shape of (640, 480) was used and the output of the binomial for the rows beyond ~540 is incorrect (shown as columns due to transpose in plotting). See figure:

![2022-11-02-cupy-incorrect-binomial-default_rng](https://user-images.githubusercontent.com/13607845/199531620-d9752982-8ceb-4942-af69-dbc8cbc9cf00.png)

When using `cupy.random.binomial` this issue isn't present, i.e., output is expected. This can be tested with tbe below script by commenting out the line `ranGen = cp.random.default_rng() # faulty` and uncommenting the line `# ranGen = cp.random # correct`, see figure for results when using `cp.random.binomial`:

![image](https://user-images.githubusercontent.com/13607845/199532141-ede3f924-c96f-4d24-b53d-0a5678729b5a.png)

### To Reproduce

```py
import cupy as cp
import numpy as np
import matplotlib.pyplot as plt
plt.rcParams["figure.autolayout"] = True

# Comment line below and uncomment the ranGen = cp.random to use a different and corret RNG
ranGen = cp.random.default_rng() # faulty
# ranGen = cp.random # correct

# Input parameters
shape = (640, 480)
p = cp.arange(np.prod(shape)).reshape(shape) / np.prod(shape)
n = 1

# Obtain binomial randoms given the probability p
res_full = ranGen.binomial(n, p) / n # Full array p at once
# Try to do it in two steps
res_half = [
    ranGen.binomial(n, p[:320]) / n,
    ranGen.binomial(n, p[320:]) / n,
]
# Stitch both arrays together
res_stitch = cp.asarray(res_half)

# Visualize the results, arrays will be transposed for plotting
fig, ax = plt.subplots(2, 3)

img = ax[0, 0].imshow(p.get().T, vmin=0, vmax=1)
fig.colorbar(img, ax=ax[0, 0])
ax[0, 0].set_title("Input probability")

# This plot (ax[0, 1]) will show the faulty output data when using ranGen =
# cp.random.default_rng()
img = ax[0, 1].imshow(res_full.get().T, vmin=0, vmax=1)
fig.colorbar(img, ax=ax[0, 1])
ax[0, 1].set_title("Random binomial: full array at once")

img = ax[1, 0].imshow(res_half[0].get().T, vmin=0, vmax=1)
fig.colorbar(img, ax=ax[1, 0])
ax[1, 0].set_title("Random binomial: first half")

img = ax[1, 1].imshow(res_half[1].get().T, vmin=0, vmax=1)
fig.colorbar(img, ax=ax[1, 1])
ax[1, 1].set_title("Random binomial: second half")

img = ax[1, 2].imshow(res_stitch.reshape(shape).get().T, vmin=0, vmax=1)
fig.colorbar(img, ax=ax[1, 2])
ax[1, 2].set_title("Random binomial: two halfs stitched")

fig.delaxes(ax[0, 2])

plt.show()
```

### Installation

Conda-Forge (`conda install ...`)

### Environment

```
>>> cp.show_config()
OS                           : Linux-6.0.5-arch1-1-x86_64-with-glibc2.36
Python Version               : 3.10.6
CuPy Version                 : 11.2.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.23.3
SciPy Version                : 1.9.1
Cython Build Version         : 0.29.32
Cython Runtime Version       : None
CUDA Root                    : /home/user/.conda/envs/python-3.10
nvcc PATH                    : None
CUDA Build Version           : 11020
CUDA Driver Version          : 11080
CUDA Runtime Version         : 11070
cuBLAS Version               : (available)
cuFFT Version                : 10702
cuRAND Version               : 10210
cuSOLVER Version             : (11, 3, 5)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 7)
Thrust Version               : 101000
CUB Build Version            : 101000
Jitify Build Version         : 343be31
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Quadro RTX 3000
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:01:00.0
```

Another conda environment (same OS).
```
>>> cp.show_config()
OS                           : Linux-6.0.5-arch1-1-x86_64-with-glibc2.10
Python Version               : 3.8.13
CuPy Version                 : 11.2.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.23.3
SciPy Version                : 1.9.1
Cython Build Version         : 0.29.32
Cython Runtime Version       : None
CUDA Root                    : /home/user/.conda/envs/python-3.8
nvcc PATH                    : None
CUDA Build Version           : 11020
CUDA Driver Version          : 11080
CUDA Runtime Version         : 11070
cuBLAS Version               : (available)
cuFFT Version                : 10702
cuRAND Version               : 10210
cuSOLVER Version             : (11, 3, 5)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 7)
Thrust Version               : 101000
CUB Build Version            : 101000
Jitify Build Version         : 343be31
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Quadro RTX 3000
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:01:00.0
```

The issue was discovered on Windows 10 (will try to upload config later), verified on linux system with kernel 5.19 (I don't remember nvidia driver version anymore, but either 515.76 or 520.56), on a linux installation in WSL on Windows 10 and all suffer from the same issue.

**Comment by emcastillo** [comment]
Thanks for all the detailed info, this may affect all the random distributions so we better look into this asap.

**Comment by Neburski** [comment]
I tried a couple of the other distributions. Both the binomial and the Poisson distribution are affected, but the others seem to not be affected (I did not extensively test them). Figure for the: binomial, Poisson, random, and standard normal distributions.

![image](https://user-images.githubusercontent.com/13607845/199688449-41062af1-3325-41f9-a2d2-47e797805a72.png)

Note the same pattern for the binomial and the Poisson distributions.
